cpp/src/utilities/cython.cu

```diff
@@ -0,0 +1,495 @@
+std::pair<size_t, weight_t> call_louvain(raft::handle_t const& handle,
+                                         graph_view_t const& graph_view,
+                                         void* identifiers,
+                                         void* parts,
+                                         size_t max_level,
+                                         weight_t resolution)
+{
+  thrust::copy(  // rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+    thrust::device,
+    thrust::make_counting_iterator(graph_view.get_local_vertex_first()),
+    thrust::make_counting_iterator(graph_view.get_local_vertex_last()),
+    reinterpret_cast<typename graph_view_t::vertex_type*>(identifiers));
+
+  return louvain(handle,
+                 graph_view,
+                 reinterpret_cast<typename graph_view_t::vertex_type*>(parts),
+                 max_level,
+                 static_cast<weight_t>(resolution));
+}
+
+}  // namespace detail
+
+namespace detail {
+
+// Final, fully-templatized call.
+template <bool transposed,
+          typename return_t,
+          typename function_t,
+          typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool is_multi_gpu>
+return_t call_function(raft::handle_t const& handle,
+                       graph_container_t const& graph_container,
+                       function_t function)
+{
+  auto graph =
+    create_graph<vertex_t, edge_t, weight_t, transposed, is_multi_gpu>(handle, graph_container);
+
+  return function(handle, graph->view());
+}
+
+// Makes another call based on vertex_t
+template <bool transposed,
+          typename return_t,
+          typename function_t,
+          typename edge_t,
+          typename weight_t,
+          bool is_multi_gpu>
+return_t call_function(raft::handle_t const& handle,
+                       graph_container_t const& graph_container,
+                       function_t function)
+{
+  // FIXME: add support for unsigned ints once the code that relies on vertex_t
+  // and edge_t being signed is retired.
+  if (graph_container.vertexType == numberTypeEnum::int32Type) {
+    return call_function<transposed, return_t, function_t, int32_t, edge_t, weight_t, is_multi_gpu>(
+      handle, graph_container, function);
+  } else if (graph_container.vertexType == numberTypeEnum::int64Type) {
+    return call_function<transposed, return_t, function_t, int64_t, edge_t, weight_t, is_multi_gpu>(
+      handle, graph_container, function);
+  } else {
+    CUGRAPH_FAIL("vertexType unsupported");
+  }
+}
+
+// Makes another call based on edge_t
+template <bool transposed,
+          typename return_t,
+          typename function_t,
+          typename weight_t,
+          bool is_multi_gpu>
+return_t call_function(raft::handle_t const& handle,
+                       graph_container_t const& graph_container,
+                       function_t function)
+{
+  if (graph_container.edgeType == numberTypeEnum::int32Type) {
```

Comment:
To avoid the creation of graph_t<int64_t, int32_t, ...> which seems like an unused/unnecessary type, you can combine the vertex_t and edge_t into a single call here.  Eliminate the call_function based on vertex_t above and do a more complex if statement here:

`if ((graph_container.vertexType == numberTypeEnum::int32Type) && (graph_container.edgeType == numberTypeEnum::int32Type)) {
  ...
`

If you only include the pairs (int32,int32), (int32,int64), (int64,int64) in the if statements then only those versions of the graph object would need to be instantiated.

Reply:
Thanks, I implemented that and it'll be in the next commit.